### PR TITLE
Fix UNDEFINED payment

### PR DESCRIPTION
**Pull request by daliwei**
cc @pierre

### CODE REVIEW

File: spec/orbital/remote/shared_examples_for_payment_flow.rb [hunk 1]
```diff
@@ -117,4 +117,34 @@
...
+    response = Killbill::Orbital::OrbitalResponse.create(:api_call => 'authorization',
+                                                         :kb_account_id => @pm.kb_account_id,
+                                                         :kb_payment_id => @kb_payment.id,
+                                                         :kb_payment_transaction_id => @kb_payment.transactions[0].id,
+                                                         :kb_tenant_id => @call_context.tenant_id,
+                                                         :message => '{"exception_message":"Timeout","payment_plugin_status":"UNDEFINED"}',
+                                                         :created_at => Time.now,
+                                                         :updated_at => Time.now)
+
+    # Set skip_gw=true, to avoid calling the report API
```
Comment (pierre) [inline spec/orbital/remote/shared_examples_for_payment_flow.rb]: Wrong comment?

File: lib/orbital/ext/active_merchant/active_merchant.rb [hunk 2]
```diff
@@ -13,7 +13,7 @@ def store(creditcard, options = {})
         # Workaround: unmask the PAN if needed
         # TODO We could call on-the-fly retrieve_customer_profile instead in PaymentPlugin#get_payment_source to
         # avoid having to store the PAN, but this requires a specific merchant account setting
-        response.params['cc_account_num'] = creditcard.number if response.params['cc_account_num'].starts_with?('XXXX')
+        response.params['cc_account_num'] = creditcard.number if response.params['cc_account_num'].include?('XXXX')
```
Comment (pierre) [inline lib/orbital/ext/active_merchant/active_merchant.rb:16]: Why this change?

File: lib/orbital/models/response.rb [hunk 3]
```diff
@@ -80,6 +80,27 @@ def self.orbital_response_params(response)
...
+          error_details = {}
+          original_message = message
+        end
+        error_details['original_message'] = original_message unless original_message.blank?
+        error_details['payment_plugin_status'] = 'CANCELED'
+
+        updated_attributes = {
+            :message => error_details.to_json,
+            :success => false,
+            :updated_at => Time.now.utc
```
Comment (pierre) [inline lib/orbital/models/response.rb]: This should use the `@clock` instead.

File: lib/orbital/api.rb [hunk 4]
```diff
@@ -74,11 +76,38 @@ def refund_payment(kb_account_id, kb_payment_id, kb_payment_transaction_id, kb_p
...
+          next unless transaction_info_plugin.status == :UNDEFINED
+
+          orbital_response_id = find_value_from_properties(transaction_info_plugin.properties, 'orbitalResponseId')
+          if orbital_response_id.nil?
+            logger.warn("Unable to fix UNDEFINED kb_transaction_id='#{transaction_info_plugin.kb_transaction_payment_id}' (orbital_response_id not specified)")
+            next
+          end
+          response = OrbitalResponse.find_by(:id => orbital_response_id)
+          if response.nil?
+            logger.warn("Unable to fix UNDEFINED kb_transaction_id='#{transaction_info_plugin.kb_transaction_payment_id}' (CyberSource response='#{cybersource_response_id}' not found)")
```
Comment (pierre) [inline lib/orbital/api.rb]: `Orbital`, not `CyberSource`.

File: lib/orbital/api.rb [hunk 5]
```diff
@@ -2,6 +2,8 @@ module Killbill #:nodoc:
   module Orbital #:nodoc:
     class PaymentPlugin < ::Killbill::Plugin::ActiveMerchant::PaymentPlugin
 
+      FIVE_MINUTES_AGO = (1 * 300)
```
Comment (pierre) [inline lib/orbital/api.rb]: How is the delay used? For CyberSource, it is used as a timeout for the reporting API. But here, the state will never be fixed.

Wouldn't it be easier to simply call [fixPaymentTransactionState](https://github.com/killbill/killbill-api/blob/master/src/main/java/org/killbill/billing/payment/api/AdminPaymentApi.java#L43) instead (from a control plugin for instance)?

**Review by pierre — CHANGES_REQUESTED** [review]

File: lib/orbital/api.rb (hunk 4 above)
Comment (daliwei) [reply lib/orbital/api.rb]: Fixed in 8ae728b

File: lib/orbital/ext/active_merchant/active_merchant.rb (hunk 2 above)
Comment (daliwei) [reply lib/orbital/ext/active_merchant/active_merchant.rb:16]: Somehow, Orbital returns the card number like `44444XXX...1111`; so the hack done in d6413fbb1a41e38002629a38ec0d790bf7047147 won't work anymore.

File: spec/orbital/remote/shared_examples_for_payment_flow.rb (hunk 1 above)
Comment (daliwei) [reply spec/orbital/remote/shared_examples_for_payment_flow.rb]: Fixed in 8ae728b

File: lib/orbital/models/response.rb (hunk 3 above)
Comment (daliwei) [reply lib/orbital/models/response.rb]: Tried to do so; how can I get `@clock` in `response.rb`?

Looks like we use `Time.now.utc` in other plugins as well, e.g, CyberSource and PayPal

File: lib/orbital/api.rb (hunk 5 above)
Comment (daliwei) [reply lib/orbital/api.rb]: Yes, the state will never be fixed. I put it in the payment plugin is for consistency with other plugins such as CyberSource. If we put it in a control plugin, that will work as well; but we would need some `if else` logic to identify if it is a `Orbital` payment and then cancel it. Seems to me that the cancelation logic would be in two places separately. For some payment gateway, you need to look at the plugin itself; while for others, you need to look at the control plugin.

Comment (pierre) [reply lib/orbital/api.rb]: It sounds like we never want to return `UNDEFINED`? Any timeout operation would result in a cancelled operation -- is that what we want?

* If indeed we never want to return `UNDEFINED`, we should just never return it in the first place (we would save on extra logic, DAO calls, etc.)?
* Otherwise, should we make it explicit that we want to cancel the operation (otherwise any GET will _fix_ it implicitly)?

Comment (daliwei) [reply lib/orbital/api.rb]: Yea. I would prefer the first option: never return `UNDEFINED` for Orbital. Sounds good to you if I go ahead to make that change?

Comment (pierre) [reply lib/orbital/api.rb]: Yup, seems easier and simpler indeed. Could you just make it a configurable property?

Comment (daliwei) [reply lib/orbital/api.rb]: Well I didn't see an easy way to make it configurable (such as consuming a property that is passed in); but I made a simple enough modification in 95b4341 to return `CANCELED` status for `UNDEFINED`.

Comment (pierre) [reply lib/orbital/api.rb]: Patch looks fine but I'm wondering if we should do this in `api.rb` instead, right before returning the transaction (when returning `authorizePayment`, `getPaymentInfo`, ...)?

That way, we could add a log entry for these originally `UNDEFINED` transactions that we are _fixing_ on the fly, if we need to go back and identify them later on (it would also make it possible to have a config flag I believe).

Not trying to be difficult but to explain where I'm coming from: we still have very little experience with Orbital, and I'm just trying to be pro-active and be ready if (when?) we experience a large outage and we have to go through these transactions manually (more data, logs, configuration knobs we have, easier it will be). But if you think I'm over-thinking it, happy to merge it! 😼

Comment (daliwei) [reply lib/orbital/api.rb]: Well, the raw messages we persisted already indicates that this is an `UNDEFINED` response. I can add a log for better tracking at `to_transaction_plguin_info`; but not sure we would need to put that logic for each call when we return `trx plugin info`?

Comment (pierre) [reply lib/orbital/api.rb]: Right, the data is in the database, but a `like` query on that table won't be scalable (full table scan) as we record more and more payments (unless we go to a warehouse, but ETLs are often delayed, so not useful during outages). That's why I'm looking for a different solution (log storage search engine, indexed by timestamp).

Ideally, we would only log when we overwrite the payment status the first time (typically in `authorizePayment`).
